packages/react-dom-bindings/src/client/ReactDOMHostConfig.js

```diff
@@ -3010,3 +3046,257 @@ export function isHostHoistableType(
+  suspendedState = {
+    stylesheets: new Map(),
+    preloadCount: 0,
+    count: 0,
+    unsuspend: null,
+  };
+}
+
+export function suspendInstance(type: Type, props: Props): void {
+  return;
+}
+
+function onPreloadComplete(this: SuspendedState) {
+  this.count--;
+  this.preloadCount--;
+  if (this.preloadCount === 0) {
+    insertSuspendedStylesheets(this, this.stylesheets);
+  }
+  unsuspendIfUnblocked(this);
+}
+
+export function suspendResource(
+  hoistableRoot: HoistableRoot,
+  resource: Resource,
+  props: any,
+): void {
+  if (suspendedState === null) {
+    throw new Error(
+      'Internal React Error: suspendedState null when it was expected to exists. Please report this as a React bug.',
+    );
+  }
+  const state = suspendedState;
+  if (resource.type === 'stylesheet' && resource.instance === null) {
+    const qualifiedProps: StylesheetQualifyingProps = props;
+    const key = getStyleKey(qualifiedProps.href);
+
+    // Attempt to hydrate instance from DOM
+    let instance: null | Instance = hoistableRoot.querySelector(
+      getStylesheetSelectorFromKey(key),
+    );
+    if (instance) {
+      // If this instance has a loading state it came from the Fizz runtime.
+      // If there is not loading state it is assumed to have been server rendered
+      // as part of the preamble and therefore synchronously loaded. It could have
+      // errored however which we still do not yet have a means to detect. For now
+      // we assume it is loaded.
+      const maybeLoadingState: ?Promise<mixed> = (instance: any)._p;
+      if (
+        maybeLoadingState !== null &&
+        typeof maybeLoadingState === 'object' &&
+        // $FlowFixMe[method-unbinding]
+        typeof maybeLoadingState.then === 'function'
+      ) {
+        const loadingState = maybeLoadingState;
+        state.count++;
+        const ping = onUnsuspend.bind(state);
+        loadingState.then(ping, ping);
+      }
+      resource.instance = instance;
+      markNodeAsHoistable(instance);
+      return;
+    }
+
+    const ownerDocument = getDocumentFromRoot(hoistableRoot);
+
+    const stylesheetProps = stylesheetPropsFromRawProps(props);
+    const preloadProps = preloadPropsMap.get(key);
+    if (preloadProps) {
+      adoptPreloadPropsForStylesheet(stylesheetProps, preloadProps);
+    }
+
+    // Construct and insert a new instance
+    instance = ownerDocument.createElement('link');
+    markNodeAsHoistable(instance);
+    const linkInstance: HTMLLinkElement = (instance: any);
+    (linkInstance: any)._p = new Promise((resolve, reject) => {
+      linkInstance.onload = resolve;
+      linkInstance.onerror = reject;
+    });
+    setInitialProperties(instance, 'link', stylesheetProps);
+    resource.instance = instance;
+    state.stylesheets.set(resource, hoistableRoot);
+
+    const preloadEl = resource.state.preload;
+    if (preloadEl && (resource.state.loading & Settled) === NotLoaded) {
+      state.count++;
+      state.preloadCount++;
```

Comment:
Could you explain why `count` and `preloadCount` are separate? As far as I can tell a resource only gets one or the other type of listener.

Reply:
My initial thinking was that we would insert stylesheet as soon as all the preloads for the stylesheet finished. Later when we support more kinds of suspensey things we'll need to decide whether we want to actually delay the stylesheet inserts until everything has finished but at the moment these are indistinguishable.

`preloadCount` is a count of the outstanding preloads for the stylesheets specifically, not a tracker for the preloading of arbitrary assets. count serves as a general purpose task counter but if we want to unlock the stylesheet insertion upon just their preloads finishing I needed to track that separately.

That said, your comment got me to rethink this logic and for now I think we should assume stylesheets will go after ALL other things are preloaded which means we can just use a single ref count. I pushed an update that removes the special case for stylesheet preloads and fixes stuff from the other comments